newton/_src/utils/schema_resolver.py

```diff
@@ -0,0 +1,601 @@
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import Any, ClassVar
+
+import warp as wp
+from pxr import Usd
+
+
+@dataclass
+class Attribute:
+    """
+    Specifies a USD attribute and its transformation function.
+
+    Args:
+        usd_name (str): The name of the USD attribute.
+        default (Any | None): Default USD-authored value from schema, if any.
+        transform (Callable[[Any], Any] | None): A function to transform the attribute value.
+    """
+
+    usd_name: str
+    default: Any | None = None
+    transform: Callable[[Any], Any] | None = None
+
+
+class schemaPlugin:
+    name: str
+    # mapping is a dictionary for known variables in Newton. Its purpose is to map usd attributes to exisiting Newton data.
+    # prim_type -> variable -> list[Attribute]
+    mapping: dict[str, dict[str, list[Attribute]]]
+
+    # extra_attr_namespaces is a list of namespaces for extra attributes that are not in the mapping.
+    extra_attr_namespaces: ClassVar[list[str]] = []
+
+    def __init__(self) -> None:
+        # Precompute the full set of USD attribute names referenced by this plugin's mapping.
+        names: set[str] = set()
+        try:
+            mapping_items = self.mapping.items()
+        except AttributeError:
+            mapping_items = []
+        for _prim_type, var_map in mapping_items:
+            try:
+                var_items = var_map.items()
+            except AttributeError:
+                continue
+            for _var, specs in var_items:
+                for spec in specs:
+                    names.add(spec.usd_name)
+        self._engine_attributes: list[str] = list(names)
+
+    def get_value(self, prim, prim_type: str, key: str) -> tuple[Any, str] | None:
+        """
+        Get attribute value for a given prim type and key.
+
+        Args:
+            prim: USD prim to query
+            prim_type: Prim type ("scene", "joint", "shape", "body", "material", "actuator")
```

Comment:
Can you add an `IntEnum` for the prim_type? It will be clearer to know what possible types exist.